### PR TITLE
[train] add minimal installation test suite

### LINKED ISSUE #22118: [train] Add minimal install test suite
We recently had to revert a PR (https://github.com/ray-project/ray/pull/22117) since it changed the behavior of Ray Train to require torch. This was only uncovered because of a datasets release test failing that did not have a torch dependency.

we should a minimal install test suite to make sure that ray train tensorflow works even without torch installed, and vice versa. Ray train should continue to work without tune or horovod installed.

### PR DESCRIPTION
**Pull request by matthewdeng**
## Why are these changes needed?

Adding a minimal test suite to catch any regressions from accidentally adding backend imports (e.g. `torch`, `tensorflow`, `horovod`) to the main import path.

**Example:** If I'm running Ray Train with `tensorflow`, I should not be required to have `torch` installed.

## Related issue number

This has caused problems twice in the past.
1. https://github.com/ray-project/ray/pull/20767 - error when `ray[tune]` is not installed.
2. https://github.com/ray-project/ray/pull/22117 - error when `torch>=1.8.1` is not installed.

The fix in [1] was intended to add a test for minimal installation by splitting the Tune tests, but did not account for the fact that the ML dependencies are still installed.

The tests in this PR explicitly only install the minimal installation and further tests that `torch`, `tensorflow`, and `horovod` are not installed.

Closes #22118 

## Checks
Manually added `import torch` to `ray/train/callbacks/__init__.py`, and verified that the test failed. The test function itself never runs as it fails when importing at the top of the file. I can move these imports directly into the test function, but this seems like a sufficient way to catch any regressions.

```
================================================================================
==================== Test output for //python/ray/train:test_minimal:
Traceback (most recent call last):
  File "/private/var/tmp/_bazel_matt/2236ba5679fc7926f9b836f43bdc5329/execroot/com_github_ray_project_ray/bazel-out/darwin-opt/bin/python/ray/train/test_minimal.runfiles/com_github_ray_project_ray/python/ray/train/tests/test_minimal.py", line 6, in <module>
    import ray.train as train
  File "/Users/matt/anaconda3/envs/min/lib/python3.7/site-packages/ray/train/__init__.py", line 2, in <module>
    from ray.train.callbacks import TrainingCallback
  File "/Users/matt/anaconda3/envs/min/lib/python3.7/site-packages/ray/train/callbacks/__init__.py", line 3, in <module>
    import torch
ModuleNotFoundError: No module named 'torch'
================================================================================
```

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: .buildkite/pipeline.ml.yml [hunk 1]
```diff
@@ -290,7 +290,7 @@
   commands:
     - cleanup() { if [ "${BUILDKITE_PULL_REQUEST}" = "false" ]; then ./ci/travis/upload_build_info.sh; fi }; trap cleanup EXIT
     - TRAIN_TESTING=1 INSTALL_HOROVOD=1 ./ci/travis/install-dependencies.sh
-    - bazel test --config=ci $(./scripts/bazel_export_options) --build_tests_only --test_tag_filters=-gpu_only,-tune python/ray/train/...
+    - bazel test --config=ci $(./scripts/bazel_export_options) --build_tests_only --test_tag_filters=-gpu_only,-minimal,-tune python/ray/train/...
```
Comment (amogkam) [inline .buildkite/pipeline.ml.yml:293]: Do we still need to have the tune label/separating tune tests vs. other train tests? The minimal install should already  catch unintended tune imports as well.

File: .buildkite/pipeline.yml [hunk 2]
```diff
@@ -394,3 +394,12 @@
...
     - ./ci/travis/install-dependencies.sh
     - bazel test --config=ci $(./scripts/bazel_export_options) --build_tests_only --test_tag_filters=soft_imports python/ray/tune/...
+
+- label: ":steam_locomotive: Train minimal install"
+  conditions: ["RAY_CI_TRAIN_AFFECTED"]
+  commands:
+      - cleanup() { if [ "${BUILDKITE_PULL_REQUEST}" = "false" ]; then ./ci/travis/upload_build_info.sh; fi }; trap cleanup EXIT
+      - ./ci/travis/install-minimal.sh
+      - ./ci/travis/env_info.sh
+      - python ./ci/travis/check_minimal_install.py
```
Comment (amogkam) [inline .buildkite/pipeline.yml:404]: We may want to consider adding a new `check_train_minimal_install.py` script for the dependencies that we care about.

Theoretically someone could add `torch` install to `install-minimal.sh` and we wouldn't catch that.

But might be out of scope for this PR.

**Review by amogkam — COMMENTED** [review]
Nice, this looks great! Just left some small comments.

File: .buildkite/pipeline.yml (hunk 2 above)
Comment (amogkam) [reply .buildkite/pipeline.yml:404]: Actually nvm the test_minimal.py already checks for this.

File: python/ray/train/tests/test_minimal.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+def test_failure():
+    """Tests that backend frameworks are not imported."""
+    with pytest.raises(ModuleNotFoundError):
+        import torch  # noqa: F401
+
+    with pytest.raises(ModuleNotFoundError):
+        import tensorflow  # noqa: F401
+
+    with pytest.raises(ModuleNotFoundError):
+        import horovod  # noqa: F401
```
Comment (amogkam) [inline python/ray/train/tests/test_minimal.py:88]: Add one for `from ray import tune`?

File: .buildkite/pipeline.ml.yml (hunk 1 above)
Comment (matthewdeng) [reply .buildkite/pipeline.ml.yml:293]: I don't have a super strong opinion here, but I think it's nice to still split it so you can run the tests in parallel and it helps isolating the problem easier if it fails in one suite but not the other.

File: python/ray/train/tests/test_minimal.py (hunk 3 above)
Comment (matthewdeng) [reply python/ray/train/tests/test_minimal.py:88]: Good point, added!

File: .buildkite/pipeline.yml (hunk 2 above)
Comment (matthewdeng) [reply .buildkite/pipeline.yml:404]: Yep both these are true. My thought process here is that `install-minimal.sh` will always be the bare minimum for Ray, so if something like `torch` or even `tune` is installed I'd believe there's a bigger problem at hand.

However, the other direction of change would be if `train` has more in its minimal installation path, in which case we should definitely look to use our own script/minimal imports.

**Review by amogkam — APPROVED** [review]
LGTM, thanks!

**Pull request merged**